### ISSUE TITLE
Custom grammar without STATEMENT field causes exception

### ISSUE THREAD

**Issue by haxtibal** [strictdoc-project/strictdoc#1823]
StrictDocs custom grammar feature makes it quite versatile. One can not only document requirements, but also design elements, threats from threat modelling, test cases, and so on.

I defined a `THREAT` tag, notably without a `STATEMENT` field as it wouldn't have meaning in that context.

```
[GRAMMAR]
ELEMENTS:
- TAG: THREAT
  FIELDS:
  - TITLE: UID
    TYPE: String
    REQUIRED: TRUE
  - TITLE: TITLE
    TYPE: String
    REQUIRED: TRUE
  - TITLE: DESCRIPTION
    TYPE: String
    REQUIRED: False
  - TITLE: SEVERITY
    TYPE: SingleChoice(Low, Medium, High)
    REQUIRED: TRUE
  - TITLE: SCORE
    TYPE: String
    REQUIRED: False
```

With such a grammar `strictdoc export` (version 0.0.55) raises an exception

```
...
  File "/home/tdmg/.local/pipx/venvs/strictdoc/lib/python3.11/site-packages/strictdoc/backend/sdoc/models/node.py", line 459, in enumerate_meta_fields
    statement_field_index = grammar_field_titles.index(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: 'STATEMENT' is not in list
```

because `SDocNode.enumerate_meta_fields` depends on the `STATEMENT` field to encode a convention that says "each field before STATEMENT is a single line field".

Would it be possible to find a different convention and remove that hard coded dependency on `STATEMENT`, to support use cases where `STATEMENT` doesn't make sense?

**Comment by stanislaw** [comment]
Hi @haxtibal. This is something that I was waiting someone to ask for, and I wanted to relax the STATEMENT-convention myself.

The main reason why this was done this way was that the Web UI has to know which field is a primary field and display it accordingly. I don't have a good general solution to this just yet, so for now I simply extended StrictDoc to auto-detect the primary field from one of `{STATEMENT, DESCRIPTION, CONTENT}`. Your example has `DESCRIPTION`, so let me know if this extended but still hard-coded solution is enough for your use case. I added your example as one of the integration test that now passes. Otherwise, let me know which other primary fields would you want to have?

You can try installing StrictDoc Pip from the latest main branch and let me know if it works better:

pip install -U --pre git+https://github.com/strictdoc-project/strictdoc.git@main

An extra note:

A more flexible solution I considered was to extend the grammar element declaration to support a `PRIMARY_FIELD` where you would simply declare which field StrictDoc should treat as the primary one, e.g., `PRIMARY_FIELD: DESCRIPTION`. I am not doing this just yet unless you have the whole variety of primary fields to support.

```
ELEMENTS:
- TAG: REQUIREMENT
  PRIMARY_FIELD: DESCRIPTION
  FIELDS:
  - TITLE: UID
    TYPE: String
```

**Comment by haxtibal** [comment]
Thank you very much for the quick response and fix.

>  let me know which other primary fields would you want to have?

We're only in the beginning of setting up StrictDoc based workflows and yet unknown types lie ahead. Your solution is certainly good for the types I currently know.

> You can try installing StrictDoc Pip from the latest main branch and let me know if it works better:
> 
> pip install -U --pre git+https://github.com/strictdoc-project/strictdoc.git@main

It works.

Minor issues:
- `SDocParsingProcessor.process_document_grammar_element` is not called if `[GRAMMAR] IMPORT_FROM_FILE:` is used. This way one can bypass the `REQUIRED: True` check. I'd assume that's unintended?
- If you define both `DESCRIPTION` and `STATEMENT`, the document validates, but only the latter of the two fields is rendered. It's a contrived case, so probably not too bad.

> An extra note:
> 
> A more flexible solution I considered was to extend the grammar element declaration to support a `PRIMARY_FIELD` where you would simply declare which field StrictDoc should treat as the primary one, e.g., `PRIMARY_FIELD: DESCRIPTION`. I am not doing this just yet unless you have the whole variety of primary fields to support.
> ```
> ELEMENTS:
> - TAG: REQUIREMENT
>   PRIMARY_FIELD: DESCRIPTION

For the general case, could there be a need for multiple content fields? What are the notable consequences when a field becomes a primary field?

**Comment by stanislaw** [comment]
> Thank you very much for the quick response and fix.
> 
> > let me know which other primary fields would you want to have?
> 
> We're only in the beginning of setting up StrictDoc based workflows and yet unknown types lie ahead. Your solution is certainly good for the types I currently know.
> 
> > You can try installing StrictDoc Pip from the latest main branch and let me know if it works better:
> > pip install -U --pre git+https://github.com/strictdoc-project/strictdoc.git@main
> 
> It works.
> 
> Minor issues:
> 
>     * `SDocParsingProcessor.process_document_grammar_element` is not called if `[GRAMMAR] IMPORT_FROM_FILE:` is used. This way one can bypass the `REQUIRED: True` check. I'd assume that's unintended?

I believe this was fixed here yesterday: https://github.com/strictdoc-project/strictdoc/pull/1831. Please try again.

>     * If you define both `DESCRIPTION` and `STATEMENT`, the document validates, but only the latter of the two fields is rendered. It's a contrived case, so probably not too bad.

Yes, this is the assumption. I would like to hit a yet another reason to advance here. The current solution seems to be good enough.

> 
> > An extra note:
> > A more flexible solution I considered was to extend the grammar element declaration to support a `PRIMARY_FIELD` where you would simply declare which field StrictDoc should treat as the primary one, e.g., `PRIMARY_FIELD: DESCRIPTION`. I am not doing this just yet unless you have the whole variety of primary fields to support.
> > ```
> > ELEMENTS:
> > - TAG: REQUIREMENT
> >   PRIMARY_FIELD: DESCRIPTION
> > ```
> 
> For the general case, could there be a need for multiple content fields? What are the notable consequences when a field becomes a primary field?

I am challenging this myself, and I cannot come up with a real use for the multiple statements case yet. Here are the original and existing reasons for ONE primary field convention:

1) The assumption was that each node should have something unique to say and that should be collected in just one field.

2) StrictDoc used to have a plain-text view for rendering nodes where STATEMENTS were simply rendered as plain text while all other fields were highlighted as fields, single- and multiline. We removed that original UI markup, so this is not actual anymore because everything is rendered as tables now and all node field names are always present.

3) The DTR screen has this branch where only a short version of a statement is rendered. Without STATEMENT/DESCRIPTION/CONTENT convention, it is not clear how would StrictDoc decide which field to render in the small requirement badge on the DTR:

```
  {%- if truncated_statement is true -%}
    {# truncated in DTR tiny card #}
    <sdoc-requirement-field data-field-label="statement">
      <sdoc-autogen>{{ view_object.markup_renderer.render_truncated_requirement_statement(requirement) }}</sdoc-autogen>
    </sdoc-requirement-field>
  {%- else -%}
```

4) The upcoming feature https://github.com/strictdoc-project/strictdoc/issues/1229 has the idea of validating requirements statements against guidelines such as INCOSE requirements guideline. StrictDoc has to have knowledge of which field is actually a requirement statement or not. If the STATEMENT/DECISION/CONTENT automatic convention is removed, then the grammar element or the validation command must specify which field shall be used for validation.

I am open to collecting strong user needs and considering meaningful solutions 😄

**Comment by haxtibal** [comment]
> I believe this was fixed here yesterday: #1831. Please try again.

Tested, works. Thanks!

> 1. The assumption was that each node should have something unique to say and that should be collected in just one field.

STATEMENT and RATIONALE are two separate fields, both say something important. Whether I'd expect to see one, both, or none depends on the specific view.

> 3. The DTR screen has this branch where only a short version of a statement is rendered. Without STATEMENT/DESCRIPTION/CONTENT convention, it is not clear how would StrictDoc decide which field to render in the small requirement badge on the DTR:

We intend to put (much) more into a content field than what's there in a traditional singular requirement statement. This will contradict the DTR brevity goal. Having only UID and TITLE in the tree view would be enough for us, I guess.

The story behind is we'll basically convert a superset of [arc42](https://arc42.org/overview) into StrictDoc. Any important product artifact would be documented with StrictDoc. StrictDoc would manage UIDs not only for requirements, but also for architecture/design elements, test cases, threats, risks (as in #1830) and so on. We'll inline tables and graphical elements (static SVGs and mermaid, for UML, C4, DFD diagrams) as content. All those elements would be unique, structured (custom), traceable and reviewed in a git/email based approval workflow.

> 4. The upcoming feature [Feature: basic requirements validation #1229](https://github.com/strictdoc-project/strictdoc/issues/1229) has the idea of validating requirements statements against guidelines such as INCOSE requirements guideline. StrictDoc has to have knowledge of which field is actually a requirement statement or not. If the STATEMENT/DECISION/CONTENT automatic convention is removed, then the grammar element or the validation command must specify which field shall be used for validation.

Our requirements would have either the form of ISO 29148 5.2.4, or alternatively that of agile user stories. Other elements like design or tests would again have very different validation rules, if any. That's not easy to hard code... Will they remain valid after #1129? Could StrictDoc support hooks where one can supply custom Python code for validation? Or at least make the validation optional?

**Comment by stanislaw** [comment]
> > 1. The assumption was that each node should have something unique to say and that should be collected in just one field.
> 
> STATEMENT and RATIONALE are two separate fields, both say something important. Whether I'd expect to see one, both, or none depends on the specific view.

This is true. These and other fields are always shown currently on all screens except DTR. There is a 50%-finished work on https://github.com/strictdoc-project/strictdoc/issues/1509 by @dahbar where it should be possible to customize which fields are shown.

> > 3. The DTR screen has this branch where only a short version of a statement is rendered. Without STATEMENT/DESCRIPTION/CONTENT convention, it is not clear how would StrictDoc decide which field to render in the small requirement badge on the DTR:
> 
> We intend to put (much) more into a content field than what's there in a traditional singular requirement statement. This will contradict the DTR brevity goal. Having only UID and TITLE in the tree view would be enough for us, I guess.

I understand and agree with this. I am not sure yet which rule the DTR should use for showing which content. Some (rare) requirements documents do not have TITLE used, which is another edge case. I myself use DTR from time to time and seeing the first sentences of a truncated statements has proven to be useful, so I think I would not remove it from being displayed on DTR just yet.

> The story behind is we'll basically convert a superset of [arc42](https://arc42.org/overview) into StrictDoc. Any important product artifact would be documented with StrictDoc. StrictDoc would manage UIDs not only for requirements, but also for architecture/design elements, test cases, threats, risks (as in #1830) and so on. We'll inline tables and graphical elements (static SVGs and mermaid, for UML, C4, DFD diagrams) as content. All those elements would be unique, structured (custom), traceable and reviewed in a git/email based approval workflow.

This all sounds like you are doing very serious work with StrictDoc which makes me happy. I was not aware about arc42 before and from looking at their website quickly, I thought they are a very nice product. If you have time, you could share what makes you migrate? If there is anything sensitive to your work, feel free to contact me via email directly.

> > 4. The upcoming feature [Feature: basic requirements validation #1229](https://github.com/strictdoc-project/strictdoc/issues/1229) has the idea of validating requirements statements against guidelines such as INCOSE requirements guideline. StrictDoc has to have knowledge of which field is actually a requirement statement or not. If the STATEMENT/DECISION/CONTENT automatic convention is removed, then the grammar element or the validation command must specify which field shall be used for validation.
> 
> Our requirements would have either the form of ISO 29148 5.2.4, or alternatively that of agile user stories. Other elements like design or tests would again have very different validation rules, if any. That's not easy to hard code... Will they remain valid after #1129? 

Yes, they will. A flexible/configurable solution must be found.

> Could StrictDoc support hooks where one can supply custom Python code for validation? Or at least make the validation optional?

It definitely could. I am not decided on a final concept.
